pkg/skaffold/deploy/status_check.go

```diff
@@ -59,49 +59,79 @@ type counter struct {
+	// TODO use an interface #4605
+	runCtx          *runcontext.RunContext
+	labeller        *DefaultLabeller
+	deadlineSeconds int
+	muteLogs        bool
+}
+
+// NewStatusChecker returns a status checker which runs checks on deployments and pods.
+func NewStatusChecker(runCtx *runcontext.RunContext, labeller *DefaultLabeller) StatusChecker {
+	return statusChecker{
+		muteLogs:        runCtx.Muted().MuteStatusCheck(),
+		runCtx:          runCtx,
+		labeller:        labeller,
+		deadlineSeconds: runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds,
+	}
+}
+
+// Run runs the status checks on deployments and pods deployed in current skaffold dev iteration.
+func (s statusChecker) Check(ctx context.Context, out io.Writer) error {
 	event.StatusCheckEventStarted()
-	errCode, err := statusCheck(ctx, defaultLabeller, runCtx, out)
+	errCode, err := s.statusCheck(ctx, out)
 	event.StatusCheckEventEnded(errCode, err)
 	return err
 }
 
-func statusCheck(ctx context.Context, defaultLabeller *DefaultLabeller, runCtx *runcontext.RunContext, out io.Writer) (proto.StatusCode, error) {
+func (s statusChecker) statusCheck(ctx context.Context, out io.Writer) (proto.StatusCode, error) {
 	client, err := pkgkubernetes.Client()
 	if err != nil {
 		return proto.StatusCode_STATUSCHECK_KUBECTL_CLIENT_FETCH_ERR, fmt.Errorf("getting Kubernetes client: %w", err)
 	}
 
 	deployments := make([]*resource.Deployment, 0)
-	for _, n := range runCtx.GetNamespaces() {
-		newDeployments, err := getDeployments(client, n, defaultLabeller,
-			getDeadline(runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds))
+	for _, n := range s.runCtx.GetNamespaces() {
+		newDeployments, err := getDeployments(client, n, s.labeller,
+			getDeadline(s.runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds))
 		if err != nil {
 			return proto.StatusCode_STATUSCHECK_DEPLOYMENT_FETCH_ERR, fmt.Errorf("could not fetch deployments: %w", err)
 		}
 		deployments = append(deployments, newDeployments...)
 	}
 
-	deadline := statusCheckMaxDeadline(runCtx.Pipeline().Deploy.StatusCheckDeadlineSeconds, deployments)
-
 	var wg sync.WaitGroup
 
 	c := newCounter(len(deployments))
 
+	sCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	for _, d := range deployments {
 		wg.Add(1)
 		go func(r *resource.Deployment) {
 			defer wg.Done()
 			// keep updating the resource status until it fails/succeeds/times out
-			pollDeploymentStatus(ctx, runCtx, r)
+			pollDeploymentStatus(sCtx, s.runCtx, r)
 			rcCopy := c.markProcessed(r.Status().Error())
-			printStatusCheckSummary(out, r, rcCopy)
+			s.printStatusCheckSummary(out, r, rcCopy)
+			// if one deployment is in error, cancel status checks for all deployments.
+			if r.Status().Error() != nil && r.StatusCode() != proto.StatusCode_STATUSCHECK_CONTEXT_CANCELLED {
+				cancel()
+			}
 		}(d)
 	}
 
 	// Retrieve pending deployments statuses
 	go func() {
-		printDeploymentStatus(ctx, out, deployments, deadline)
+		s.printDeploymentStatus(sCtx, out, deployments)
```

Comment:
Since this doesn't add to the WaitGroup, it's possible this code is called after `sCtx` has been cancelled.

Reply:
no it should not happen. This WaitGroup is way to wait for all the `pollDeployments to finish`
There won't be go routine leak by not adding to wait group. Having `printDeploymentStatus` run in child context of parent should gaurantee this will get cancelled.